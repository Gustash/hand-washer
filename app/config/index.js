import local from './local'
import development from './development'
import production from './production'
import colors from './colors'
import store from './store'

let config = {
  colors,
  storybookEnabled: false,
}

if (__DEV__) {
  config = {
    ...config,
    ...development,
    ...local,
    ...store,
  }
} else {
  config = {
    ...config,
    ...local,
    ...production,
    ...store,
  }
}

export { colors }
export default config
