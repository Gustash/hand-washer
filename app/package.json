{
  "name": "app",
  "//": "This allows us to import from e.g. `app/components/Button` from anywhere"
}
