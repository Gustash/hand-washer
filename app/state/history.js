import { createSlice } from '@reduxjs/toolkit'
import moment from 'moment'

const initialState = {
  storeWashes: true,
  washes: [],
}

export default createSlice({
  name: 'history',
  initialState,
  reducers: {
    logWash(state, action) {
      if (!state.storeWashes) {
        return {
          ...state,
          washes: [],
        }
      }

      const { format = 'DD-MM-YYYY HH:mm', date, ...entry } = action.payload
      const momentDate = moment(date, format)

      if (!momentDate.isValid()) {
        return state
      }

      // Get the biggest existing id and increment it to create a new unique one
      const newId = Math.max(0, ...state.washes.map(({ id }) => id)) + 1

      return {
        ...state,
        washes: [
          ...state.washes,
          {
            ...entry,
            id: newId,
            date: momentDate.format('DD-MM-YYYY HH:mm'),
          },
        ],
      }
    },
    setStoreWashes(state, action) {
      return {
        ...state,
        storeWashes: action.payload,
        washes: action.payload ? state.washes : [],
      }
    },
  },
})
