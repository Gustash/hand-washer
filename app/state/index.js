export { actions } from './slices'
export { default as rootReducer } from './reducer'
