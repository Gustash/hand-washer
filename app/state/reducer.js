import { persistReducer } from 'redux-persist'
import AsyncStorage from '@react-native-community/async-storage'
import { combineReducers } from '@reduxjs/toolkit'
import { reducers } from './slices'

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
}

const rootReducer = combineReducers(reducers)

export default persistReducer(persistConfig, rootReducer)
