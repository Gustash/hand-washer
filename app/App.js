import 'react-native-gesture-handler'
import React from 'react'
import { StyleSheet, StatusBar } from 'react-native'
import { Provider } from 'react-redux'
import { PersistGate } from 'redux-persist/integration/react'
import { NavigationContainer } from '@react-navigation/native'
import { createStackNavigator } from '@react-navigation/stack'
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5'

import Home from './screens/Home'
import WashTimer from './screens/WashTimer'
import Settings from './screens/Settings'
import config, { colors } from './config'

const Stack = createStackNavigator()

const { store, persistor } = config.createStore()

const App = () => (
  <Provider store={store}>
    <PersistGate loading={null} persistor={persistor}>
      <NavigationContainer>
        <StatusBar backgroundColor={colors.green} />
        <Stack.Navigator
          screenOptions={{
            headerStyle: styles.header,
            headerBackImage: () => (
              <FontAwesome5
                name="chevron-left"
                size={16}
                color={colors.white}
                style={styles.headerBackIcon}
              />
            ),
            headerTitleStyle: styles.headerTitle,
            headerRightContainerStyle: styles.headerRightIcon,
            headerBackTitle: 'Back',
            headerBackTitleStyle: [styles.headerTitle, styles.headerBackTitle],
          }}
        >
          <Stack.Screen
            name="Home"
            component={Home}
            options={{ title: 'Wash Timer' }}
          />
          <Stack.Screen name="Settings" component={Settings} />
          <Stack.Screen name="Timer" component={WashTimer} />
        </Stack.Navigator>
      </NavigationContainer>
    </PersistGate>
  </Provider>
)

const styles = StyleSheet.create({
  header: {
    backgroundColor: colors.green,
  },
  headerTitle: {
    color: colors.white,
    fontSize: 17,
  },
  headerRightIcon: {
    marginRight: 16,
  },
  headerBackTitle: {
    fontSize: 16,
  },
  headerBackIcon: {
    marginLeft: 16,
    marginRight: 4,
  },
})

export default App
