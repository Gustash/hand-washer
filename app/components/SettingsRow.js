import React from 'react'
import { TouchableOpacity, Text, StyleSheet } from 'react-native'
import { colors } from 'app/config'

const SettingsRow = ({ text, rightElement, style, onPress }) => (
  <TouchableOpacity
    style={[styles.container, style]}
    onPress={onPress}
    disabled={!onPress}
  >
    <Text testId="settings-row-text" style={styles.text}>
      {text}
    </Text>
    {rightElement}
  </TouchableOpacity>
)

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    borderBottomColor: colors.separator,
    borderBottomWidth: 1,
    paddingVertical: 12,
    paddingLeft: 8,
    paddingRight: 27,
    alignItems: 'center',
  },
  text: {
    color: colors.darkGrey,
    fontSize: 17,
    flex: 1,
  },
})

export default SettingsRow
