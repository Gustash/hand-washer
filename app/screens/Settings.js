import React from 'react'
import {
  SafeAreaView,
  ScrollView,
  StyleSheet,
  Switch,
  Alert,
} from 'react-native'
import { useDispatch, useSelector } from 'react-redux'

import SettingsRow from 'app/components/SettingsRow'
import { colors } from 'app/config'
import { actions } from 'app/state'

const Settings = () => {
  const dispatch = useDispatch()

  const storeWashes = useSelector(state => state.history.storeWashes)
  const shiftReminders = useSelector(state => state.shifts.shiftReminders)

  return (
    <SafeAreaView style={styles.container}>
      <ScrollView contentContainerStyle={styles.scrollViewContent}>
        <SettingsRow
          text="Store History"
          rightElement={
            <Switch
              value={storeWashes}
              onValueChange={() => {
                if (storeWashes) {
                  Alert.alert(
                    'Are you sure?',
                    'Disabling this option will clear your current history and the app will stop tracking history.',
                    [
                      {
                        text: 'No, take me back!',
                        style: 'cancel',
                      },
                      {
                        text: "I'm sure",
                        onPress: () => {
                          dispatch(actions.setStoreWashes(false))
                        },
                      },
                    ]
                  )
                  return
                }

                dispatch(actions.setStoreWashes(true))
              }}
            />
          }
          style={styles.row}
        />
        <SettingsRow
          text="Two Hour On Shift Reminders"
          rightElement={
            <Switch
              value={shiftReminders}
              onValueChange={() => {
                dispatch(actions.setShiftReminders(!shiftReminders))
              }}
            />
          }
          style={styles.row}
        />
      </ScrollView>
    </SafeAreaView>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  scrollViewContent: {
    paddingTop: 36,
    flexGrow: 1,
  },
  row: {
    marginLeft: 13,
  },
  createdByContainer: {
    alignItems: 'center',
  },
  createdByText: {
    fontSize: 17,
    color: colors.brandedText,
  },
  image: {
    marginTop: 14,
    width: 178,
    height: 40,
  },
})

export default Settings
