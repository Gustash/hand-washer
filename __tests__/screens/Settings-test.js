import React from 'react'
import { Alert } from 'react-native'
import { Provider, useSelector, useDispatch } from 'react-redux'
import configureStore from 'redux-mock-store'
import TestRenderer from 'react-test-renderer'

import Settings from '../../app/screens/Settings'
import { actions } from '../../app/state'
import SettingsRow from '../../app/components/SettingsRow'

const mockStore = configureStore([])

const mockDispatch = jest.fn()

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}))

describe('<Settings />', () => {
  let store
  let wrapper

  beforeEach(() => {
    mockDispatch.mockReset()

    store = mockStore({
      history: {
        storeWashes: true,
        washes: [],
      },
      shifts: {
        shiftActive: true,
        shiftReminders: true,
      },
    })

    useSelector.mockImplementation(selector => selector(store.getState()))
    useDispatch.mockImplementation(() => mockDispatch)

    wrapper = TestRenderer.create(
      <Provider store={store}>
        <Settings />
      </Provider>
    )
  })

  it('matches snapshot', () => {
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('handles toggling store history', () => {
    store = mockStore({
      history: {
        storeWashes: false,
        washes: [],
      },
      shifts: {
        shiftActive: true,
        shiftReminders: true,
      },
    })

    wrapper = TestRenderer.create(
      <Provider store={store}>
        <Settings />
      </Provider>
    )

    let row = wrapper.root
      .findAllByType(SettingsRow)
      .find(({ props }) => props.text === 'Store History')

    expect(row.props.rightElement.props.value).toBe(false)

    row.props.rightElement.props.onValueChange()

    expect(mockDispatch).toHaveBeenCalledTimes(1)
    expect(mockDispatch).toHaveBeenCalledWith(actions.setStoreWashes(true))

    store = mockStore({
      history: {
        storeWashes: true,
        washes: [],
      },
      shifts: {
        shiftActive: true,
        shiftReminders: true,
      },
    })

    wrapper = TestRenderer.create(
      <Provider store={store}>
        <Settings />
      </Provider>
    )

    row = wrapper.root
      .findAllByType(SettingsRow)
      .find(({ props }) => props.text === 'Store History')

    expect(row.props.rightElement.props.value).toBe(true)

    let alertButtons
    jest.spyOn(Alert, 'alert').mockImplementation((title, message, buttons) => {
      alertButtons = buttons
    })

    row.props.rightElement.props.onValueChange()

    expect(Alert.alert).toHaveBeenCalledWith(
      'Are you sure?',
      'Disabling this option will clear your current history and the app will stop tracking history.',
      [
        {
          text: 'No, take me back!',
          style: 'cancel',
        },
        {
          text: "I'm sure",
          onPress: expect.any(Function),
        },
      ]
    )

    alertButtons[1].onPress()

    expect(mockDispatch).toHaveBeenCalledTimes(2)
    expect(mockDispatch).toHaveBeenCalledWith(actions.setStoreWashes(false))
  })

  it('handles toggling shift reminders', () => {
    store = mockStore({
      history: {
        storeWashes: false,
        washes: [],
      },
      shifts: {
        shiftActive: true,
        shiftReminders: true,
      },
    })

    wrapper = TestRenderer.create(
      <Provider store={store}>
        <Settings />
      </Provider>
    )

    let row = wrapper.root
      .findAllByType(SettingsRow)
      .find(({ props }) => props.text === 'Two Hour On Shift Reminders')

    expect(row.props.rightElement.props.value).toBe(true)

    row.props.rightElement.props.onValueChange()

    expect(mockDispatch).toHaveBeenCalledTimes(1)
    expect(mockDispatch).toHaveBeenCalledWith(actions.setShiftReminders(false))

    store = mockStore({
      history: {
        storeWashes: true,
        washes: [],
      },
      shifts: {
        shiftActive: true,
        shiftReminders: false,
      },
    })

    wrapper = TestRenderer.create(
      <Provider store={store}>
        <Settings />
      </Provider>
    )

    row = wrapper.root
      .findAllByType(SettingsRow)
      .find(({ props }) => props.text === 'Two Hour On Shift Reminders')

    expect(row.props.rightElement.props.value).toBe(false)

    row.props.rightElement.props.onValueChange()

    expect(mockDispatch).toHaveBeenCalledTimes(2)
    expect(mockDispatch).toHaveBeenCalledWith(actions.setShiftReminders(true))
  })
})
