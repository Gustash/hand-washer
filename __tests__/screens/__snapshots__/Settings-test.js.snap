// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> matches snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flexGrow": 1,
        "paddingTop": 36,
      }
    }
  >
    <View>
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#E6E6E6",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "marginLeft": 13,
            "opacity": 1,
            "paddingLeft": 8,
            "paddingRight": 27,
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4A4A4A",
              "flex": 1,
              "fontSize": 17,
            }
          }
          testId="settings-row-text"
        >
          Store History
        </Text>
        <RCTSwitch
          accessibilityRole="switch"
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "height": 31,
              "width": 51,
            }
          }
          value={true}
        />
      </View>
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#E6E6E6",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "marginLeft": 13,
            "opacity": 1,
            "paddingLeft": 8,
            "paddingRight": 27,
            "paddingVertical": 12,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4A4A4A",
              "flex": 1,
              "fontSize": 17,
            }
          }
          testId="settings-row-text"
        >
          Two Hour On Shift Reminders
        </Text>
        <RCTSwitch
          accessibilityRole="switch"
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "height": 31,
              "width": 51,
            }
          }
          value={true}
        />
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
